sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/ForeachWriterTable.scala

```diff
@@ -141,17 +143,33 @@ class ForeachDataWriter[T](
       writer.process(rowConverter(record))
     } catch {
       case t: Throwable =>
-        writer.close(t)
+        closeWriter(t)
```

Comment:
so we close the writer twice: once when write fails, once when call `abort`. Instead of having the `private var closeCalled: Boolean = false`, shall we have a `private var ex: Throwable = null` and only close the writer in `abort`?

Reply:
It could be moved to `abort`, but we can't remove `closeCalled` since `commit` method could also throw exception and `abort` will be called finally. We shouldn't call `writer.close()` twice in any case, as python app will hang (that's what I fixed in this PR).